extensions/data-plane/data-plane-http/src/main/java/org/eclipse/dataspaceconnector/dataplane/http/pipeline/HttpDataSink.java

```diff
@@ -65,6 +83,42 @@ protected StatusResult<Void> transferParts(List<DataSource.Part> parts) {
     private HttpDataSink() {
     }
 
+    private interface HttpDataSinkRequest {
+        Optional<Request> makeRequestForPart(Request.Builder requestBuilder, DataSource.Part part);
+    }
+
+    private class HttpDataSinkRequestPut implements HttpDataSinkRequest {
+        @Override
+        public Optional<Request> makeRequestForPart(Request.Builder requestBuilder, DataSource.Part part) {
+            RequestBody body;
+
+            try (InputStream stream = part.openStream()) {
+                body = RequestBody.create(stream.readAllBytes());
+            } catch (IOException e) {
+                monitor.severe(format("Error reading bytes for HTTP part data %s", part.name()));
+                return Optional.empty();
+            }
+
+            return Optional.of(
+                    requestBuilder
+                            .url(endpoint)
+                            .put(body)
+                            .build());
+        }
+    }
+
+    private class HttpDataSinkRequestPost implements HttpDataSinkRequest {
```

Comment:
This and `HttpDataSinkRequestPut` could be the same thing, as the method can be passed to the request builder with the `method` method that would need the `ContentType` (that could be a parameter of `HttpDataAddress` that's  `OCTET_STREAM` by default)